docs/Migrations.md

```diff
@@ -46,4 +46,112 @@ Migrations which are Node modules should export a function which will be called
 
-You should return a promise so pinejs can wait for completion/errors.
\ No newline at end of file
+You should return a promise so pinejs can wait for completion/errors.
+
+
+## Async Migrations
+
+Migrations that may lock tables for a long time, eg. column renames on large tables, can be run as async migrations.
+
+Async migrations are executed in small batches within individual transactions and run concurrently while Pine is serving requests. Async migrations are multi-instance safe. They are synchronized between instances via the `migration status` database table and the minimum interval between two migration executions is guaranteed across instances.
+
+When Pine starts and determines there are multiple async migrations to be run, it starts all migrations concurrently. These migrations however do not run in parallel: only one migration batch is executed at any given time, then another, and so on. Thus, migrations should not depend on the result a previous one.
+Once an async migration starts, it keeps running forever even after Pine instance restarts. The async migration will stop being executed when finalized with a dedicated flag in the async migration definition. This is to guarantee that new data inserted or updated at runtime can also be migrated. The status of running migrations can be checked in `migration status` table. The configured execution parameters and the execution metrics are stored and updated after each migration batch completes. 
+
+Each async migration needs to specify a pair of an async and a sync migration part, so that the sync migration statement closes the async migration and guarantees database consistency.
+
+Async migrations are stored in the migrations folder, alongside synchronous migrations. Their file names are significant and must contain the string `.async.` so they are treated as async migrations. They can only be Typescript / Javascript (with `.ts`/`.js` extensions) files.
+
+The async migration query must have a `LIMIT` statement to limit the maximum number of affected rows per batch.
+
+
+### Async migration procedure
+An async migration procedure for adding / renaming a column may follow these steps:
+1. Deployment
+    - Add new column (with independent sync migration) to contain new data and add code accessing the new column
+	- Old code reads/writes data to/from old column and async migration migrates data from old column to new column
+	- New code reads/writes to/from new column
+	- Async migrator runs forever
+2. Deployment
+	- Finalize async migration => only sync migration part gets executed
+	- Sync migration migrates all left over data from old column to new column
+	- Sync migration deletes old column
+	- Only new code with reads/writes to/from new column 
+	- Disable all old code that reads/writes to old column
```

Comment:
In a multi-instance zero-downtime scenario, I think we should do the following (following how we do it today):
```suggestion
* Deployment 1
	- Add new column (with independent sync migration) to contain new data and add code accessing the new column.
	- Update the service's implementation to set both the old & new column on each write.
	- The service's implementation should only read the old column since the async migration still migrates data from old column to new column.
	- Async migrator runs forever.
* Deployment 2
	- Finalize async migration => only sync migration part gets executed.
	- Sync migration migrates all left over data from old column to new column.
	- Update the service's implementation to only read the new column, but still write the old one as well.
* Deployment 3
	- Update the service's implementation to stop settings the old column and remove it from the sbvr.
	- Make the old field NULLable if it isn't.
* Deployment 4
	- Delete the old column with a sync migration.
```

Reply:
@thgreasi great thx!